experiments/process_sandbox/filetree.h

```diff
@@ -0,0 +1,319 @@
+      struct DirHandle : public platform::Handle
+      {
+        using platform::Handle::Handle;
+        DirHandle(Handle&& h) : Handle(std::move(h)) {}
+      };
+
+      /**
+       * A pointer to the child directory.  This could probably be a unique
+       * pointer as it's currently used, but `std::variant` can't contain
+       * references and so we'd have to use bare pointers during traversal.
+       */
+      using DirPtr = std::shared_ptr<ExportedDirectory>;
+
+      /**
+       * The contents of this virtual directory.  This maps from file names to
+       * another instance of this class or a file / directory handle.
+       */
+      std::
+        unordered_map<std::string, std::variant<DirPtr, FileHandle, DirHandle>>
+          directory;
+
+      /**
+       * The result of a directory lookup.  This is either a handle to a real
+       * directory, or a pointer to another instance of this class.
+       */
+      using DirLookupResult =
+        std::variant<std::monostate, platform::handle_t, DirPtr>;
+
+      /**
+       * Look up a directory (either a real directory handle or a pointer to an
+       * instance of this class).  The variant contains `std::monostate` if
+       * the path is not a directory.
+       */
+      DirLookupResult get_dir(const std::string& dir)
+      {
+        auto it = directory.find(dir);
+        if (it == directory.end())
+        {
+          return {};
+        }
+        return std::visit(
+          [](auto&& arg) -> DirLookupResult {
+            using T = std::decay_t<decltype(arg)>;
+            if constexpr (std::is_same_v<T, FileHandle>)
+            {
+              return {};
+            }
+            else if constexpr (std::is_same_v<T, DirPtr>)
+            {
+              return arg;
+            }
+            else if constexpr (std::is_same_v<T, DirHandle>)
+            {
+              return arg.fd;
+            }
+          },
+          it->second);
+      }
+
+      /**
+       * Look up a file in this directory.  Returns the handle to the file if
+       * it exists.
+       */
+      std::optional<platform::handle_t> get_file(const std::string& file)
+      {
+        auto it = directory.find(file);
+        if (it == directory.end())
+        {
+          return {};
+        }
+        if (auto* h = std::get_if<FileHandle>(&it->second))
+        {
+          return h->fd;
+        }
+        return {};
+      }
+
+      /**
+       * Add a child directory tree to this directory.
+       */
+      void add_tree(const std::string& name, DirPtr& tree)
+      {
+        directory[name] = tree;
+      }
+
+      /**
+       * Add a directory tree represented by a directory handle.  Takes
+       * ownership of the handle.
+       */
+      void add_directory(const std::string& name, platform::Handle&& h)
+      {
+        directory[name] = DirHandle(std::move(h));
+      }
+
+      /**
+       * Add a file, represented by a directory handle.  Takes ownership of the
```

Comment:
"directory handle"?

Reply:
You removed the `*` comment on your change request. But yeah, that looks good, thanks!